plugins/com.google.cloud.tools.eclipse.appengine.deploy/src/com/google/cloud/tools/eclipse/appengine/deploy/AppEngineStandardDeployCommandHandler.java

```diff
@@ -3,17 +3,94 @@
+import org.eclipse.wst.common.project.facet.core.IFacetedProject;
 
+import com.google.common.annotations.VisibleForTesting;
+
+/**
+ * Command handler to deploy an App Engine web application project to App Engine Standard.
+ * <p>
+ * It copies the project's exploded WAR to a staging directory and then executes staging and deploy operations
+ * provided by the App Engine Plugins Core Library.
+ */
 public class AppEngineStandardDeployCommandHandler extends AbstractHandler {
 
+  private ProjectToStagingExporter projectToStagingExporter;
+  private StagingDirectoryProvider stagingDirectoryProvider;
+  private FacetedProjectHelper facetedProjectHelper;
+  
+  public AppEngineStandardDeployCommandHandler() {
+    this(new ProjectToStagingExporter(), new DialogStagingDirectoryProvider(), new FacetedProjectHelper());
+  }
+  
+  @VisibleForTesting
+  AppEngineStandardDeployCommandHandler(ProjectToStagingExporter exporter, 
+                                        StagingDirectoryProvider stagingDirectoryProvider,
+                                        FacetedProjectHelper facetedProjectHelper) {
+    if (exporter == null) {
+      projectToStagingExporter = new ProjectToStagingExporter();
+    } else {
+      projectToStagingExporter = exporter;
+    }
+    if (stagingDirectoryProvider == null) {
+      this.stagingDirectoryProvider = new DialogStagingDirectoryProvider();
+    } else {
+      this.stagingDirectoryProvider = stagingDirectoryProvider;
+    }
+    if (facetedProjectHelper == null) {
+      this.facetedProjectHelper = new FacetedProjectHelper();
+    } else {
+      this.facetedProjectHelper = facetedProjectHelper;
+    }
+  }
+  
   @Override
   public Object execute(ExecutionEvent event) throws ExecutionException {
-    new MessageDialog(HandlerUtil.getActiveShell(event), Messages.getString("deploy.standard.dialog.title"), null, //$NON-NLS-1$
-      Messages.getString("deploy.standard.dialog.message"), //$NON-NLS-1$
-      MessageDialog.INFORMATION, new String[]{Messages.getString("button.ok.label")}, 0).open(); //$NON-NLS-1$
-    return null;
+    try {
+      IEvaluationContext applicationContext = (IEvaluationContext)event.getApplicationContext();
+      //TODO should we use the approach from http://www.programcreek.com/2013/02/eclipse-rcp-tutorial-add-a-menu-item-to-the-popup-menu-when-right-click-a-file/
+      IStructuredSelection selection = 
+          (IStructuredSelection) applicationContext.getVariable(ISources.ACTIVE_CURRENT_SELECTION_NAME);
+      if (selection.getFirstElement() instanceof IProject) {
+        IProject project = (IProject) selection.getFirstElement();
+        IFacetedProject facetedProject = facetedProjectHelper.getFacetedProject(project);
+        // TODO replace with constant from AppEngineFacet (after possibly relocating that class)
+        if (facetedProjectHelper.projectHasFacet(facetedProject, 
+                                                 "com.google.cloud.tools.eclipse.appengine.facet")) {
```

Comment:
Don't we have a constant somewhere for this string?  We reference it from multiple places.